src/viam/app/app_client.py

```diff
@@ -0,0 +1,791 @@
+                filter).
+            dest (Optional[str]): Optional filepath to write the log entries to.
+            errors_only (Optional[bool]): Optional boolean specifying whether or not to only include error logs. Defaults to True.
+            num_pages: Optional number of pages of logs to return. Passing 0 returns all pages. Defaults to 1. All pages or the first
+                `num_pages` pages will be returned, whichever comes first.
+
+        Raises:
+            GRPCError: If an invalid robot part ID is passed.
+            AssertionError: If `num_pages` is less than 0.
+
+        Returns:
+            List[viam.app.app_client.LogEntry]: The list of log entries.
+        """
+        num_pages = num_pages if num_pages is not None else 1
+        assert num_pages >= 0
+        errors_only = errors_only if errors_only else True
+        page_token = ""
+        logs = []
+
+        if num_pages == 0:
+            while True:
+                next_page_token = await self._get_robot_part_logs(
+                    robot_part_id=robot_part_id, filter=filter, errors_only=errors_only, page_token=page_token, logs=logs
+                )
+                if not next_page_token or len(next_page_token) == 0:
+                    break
+                page_token = next_page_token
+        else:
+            while True:
+                next_page_token = await self._get_robot_part_logs(
+                    robot_part_id=robot_part_id, filter=filter, errors_only=errors_only, page_token=page_token, logs=logs
+                )
+                num_pages -= 1
+                if (not next_page_token or len(next_page_token) == 0) or num_pages == 0:
+                    break
+                page_token = next_page_token
+
+        if dest:
+            try:
+                file = open(dest, "w")
+                for log in logs:
+                    time = log.time
+                    level = log.level.upper()
+                    logger_name = log.logger_name.split(".")[0]
+                    file_name = log.caller["File"]
+                    line_number = int(log.caller["Line"])
+                    message = log.message
+                    file.write(f"{time}\t\t{level}\t{logger_name}\t{file_name}:{line_number}\t{message}\n")
+            except Exception as e:
+                LOGGER.error(f"Failed to write robot part logs to file {dest}", exc_info=e)
+
+        return logs
+
+    async def _get_robot_part_logs(
+        self, robot_part_id: str, filter: str, errors_only: bool, page_token: str, logs: List[LogEntryPB]
+    ) -> str:
+        request = GetRobotPartLogsRequest(id=robot_part_id, errors_only=errors_only, filter=filter, page_token=page_token)
+        response: GetRobotPartLogsResponse = await self._location_client.GetRobotPartLogs(request, metadata=self._metadata)
+        logs += [LogEntry.from_proto(log) for log in response.logs]
```

Comment:
I suspect this is python idiomatic so no need to change anything, but: let the record show that passing a mutable data structure to a function, mutating it, and returning some other value makes me sad :(

Reply:
Lolz

Reply:
I agree, I dislike this as well. I don't think this should be an inout param. I think this function should return a tuple of `(List[LogEntry], str)`, and the calling function should do what it wants with the list of logs